Feature/user panel upload

ThomasFinnern opened the pull request:
This feature adds a user panel view as menu selection on the frontend

**Actual menu selections** are:
* userpanel
* userupload
* usercategories
* userimages

From the entry point, the others can be reached via buttons

**Edit possibilities**
* usercategory
* userimage

**WIP**
The actual code **can be tested**. The code it is merged from is used on some websites. 
Expected further changes are a general code formatting and cleaning

Attention: The version number 4.2.0.**X**  has a trailing number indicating the version of next part updates. 
The code will iterate over some steps and with increasing X we can trace the error back to the correct code version.
At the finish line i hope to not forget to remove this number

**Tasks**
There have to be several decisions:

* Which menu start points for frontend do we support
* Which of the following templates  are still necessare 
  categories, category, categoryform, gallery, image, imageform, images
* .
* .

**Next Steps**
* The code may be tested by you.
* I Thomas will look into each changed file, correct comments, todos and tell when the files are finished

ThomasFinnern commented on site/com_joomgallery/src/Model/GalleryModel.php:
```diff
@@ -201,7 +201,7 @@ protected function setImagesModelState(ListModel &$listModel, array $fields = ar
       // Get query variables sent by the images form
       $imgform_list = $this->app->input->get('list', array());
-      $imgform_limitstart = $this->app->getUserStateFromRequest('joom.galleryview.limitstart', 'limitstart', 0, 'uint');
+      $imgform_limitstart = $this->app->getInput()->get('limitstart', 0, 'int');
```
i chaged it back now

ThomasFinnern commented on site/com_joomgallery/src/Service/DefaultRouter.php:
```diff
@@ -103,30 +103,59 @@ public function __construct(SiteApplication $app, AbstractMenu $menu, ?CategoryF
+    $this->registerView($images);
 
     $image = new RouterViewConfiguration('image');
-    $image->setKey('id')->setParent($images);
```
Removed

ThomasFinnern commented on site/com_joomgallery/src/Service/DefaultRouter.php:
```diff
@@ -1,98 +1,98 @@
+  public function __construct(SiteApplication $app, AbstractMenu $menu, ?CategoryFactoryInterface $categoryFactory, DatabaseInterface $db, $skipSelf = false)
+  {
     // Get router config value
+    // ToDO: why not         $this->noIDs = (bool) $params->get('sef_ids'); or combination ?
```
Removed comment
In global configuration joomla has its own parameter. I see now it is called 'sef' 
My question is : Why do we need a own parameter. I assume it was introduced when J1/joomgallery SEF was not stable enough

ThomasFinnern replied:
Where it does come from was clear (see your comment above) My question was more "why do we enable the user to have different url systems, side by side" you may tell me when we talk next time

MrMusic commented:
@ThomasFinnern Thank you very much for the PR. It looks very good!
The biggest issue is that image upload is not working. When attempting to upload, the following message appears:
```
Uncaught (in promise) SyntaxError: JSON.parse: unterminated string at line 1 column 12 of the JSON data
sendFetchRequest jgprocessor.js:335
```
in the console.

reilldesign commented:
@ThomasFinnern Great work. Many thanks for this great pull request. I have already had the opportunity to test it extensively.  @MrMusic But I also get the Javascript error message when uploading.


`Uncaught (in promise) SyntaxError: JSON.parse: unterminated string at line 1 column 12 of the JSON data
    sendFetchRequest jgprocessor.js:335
    startSaveRequest jgprocessor.js:407
    emitAll index.js:131
    emit index.js:33
    emit Uppy.js:312
    node_modules uppy-uploader.js:17167
    _emitSuccess upload.js:468
    _resumeUpload upload.js:632
    promise callback*./node_modules/tus-js-client/lib.esm/upload.js/_resumeUpload/< upload.js:627
    promise callback*_resumeUpload upload.js:586
    _startSingleUpload upload.js:358
    start upload.js:219
    promise callback*start upload.js:192
    qRequest index.js:424
    _call2 RateLimitedQueue.js:231
    run RateLimitedQueue.js:113
    node_modules uppy-uploader.js:17280
    _uploadLocalFile2 index.js:229
    node_modules uppy-uploader.js:17372
    _uploadFiles2 index.js:514
    value index.js:122
    _runUpload2 Uppy.js:1726
    node_modules uppy-uploader.js:8312
    promise callback*upload Uppy.js:1225
    node_modules uppy-uploader.js:15765
    Preact 27
    <anonymous> UIPlugin.js:101
`

MrMusic commented:
Currently, the 'User Panel' is almost identical to the 'User images'. Therefore, I would recommend redesigning it slightly:
I would only display a small statistic (number of categories, number of images...) in the 'User Panel', similar to the backend.
If the 'Upload limits' (Max. User categories, Max. user images, etc.) are integrated, these limits should also be displayed in the 'User Panel'.
Perhaps some user data from Joomla! could also be displayed (username, profile image, etc.)?
However, I would remove the list of images from the 'User Panel'.
Any other opinions?

ThomasFinnern commented:
Actually i am improving the panel view: Show latest categories / images in a card box area

eumel1602 commented:
Bei "User image edit"

**Reiter Images.**
Hier dreht sich das joomla zeichen immer ich kann kein thumb , original oder detail bild anschauen (Bild 1)
**Reiter Publishing:**
Ich kann Approved "Yes" nicht ändern, die auswahl gibt es (Bild 2)
**Reiter  Metadata:**
- Man kann bei einem bild keine Metadata eintragen, bearbeiten. nach dem speichern alles wieder weg

![1](https://github.com/user-attachments/assets/ebde368e-46ca-47cb-a731-ba51bb07a589)
![2](https://github.com/user-attachments/assets/5092acaf-a506-4778-a7da-8d7e72450e64)

MrMusic commented:
Ich denke approved=genehmigt sollte im frontend nicht änderbar sein. Das ist etwas was der admin im backend machen sollte.

ThomasFinnern commented:
> Ich denke approved=genehmigt sollte im frontend nicht änderbar sein. Das ist etwas was der admin im backend machen sollte.
@Elfangor93 
Wir haben noch keinen Ablauf. Daher ist nicht klar was gesetzt sein soll oder ob es änderbar sein darf

eumel1602 commented:
> Bei "User image edit" (ich möchte im FE ein bild editieren)
> **Reiter Images.** Hier dreht sich das joomla zeichen immer ich kann kein thumb , original oder detail bild anschauen (Bild 1) **Reiter Publishing:** Ich kann Approved "Yes" nicht ändern, die auswahl gibt es (Bild 2) **Reiter Metadata:**
>     * Man kann bei einem bild keine Metadata eintragen, bearbeiten. nach dem speichern alles wieder weg

Bis auf das "drehende Joomla Zeichen" sind meine drei gemeldeten Fehler behoben wurden! DANKE. 

**User images edit:**
Reiter DEtails -> ACCESS -> ich kann die verschiedneen Level z.b. auf  "special" ändern. sie werden auch übernommen aber es kommt oben eine Fehlermeldung: "You don't have permission to access this view."

Mache ich das z.b. bei den Kategroien, kommt keine Fehlermeldung. Allerdings, kann ich bei den Kategroien das Level "Super user" festlegen, und sperre mich dadurch selber aus und komme nie wieder ran an die Kategrie

eumel1602 commented:
setzte ich ein passwort für eine kategorie und gehe dann wieder in die Einstellungen, fehlt die Sprachverknüpfung... 
![1](https://github.com/user-attachments/assets/1d5e710a-3370-4255-90a6-cfe75a02f390)

Ich konnte das Passwort auch nicht wieder löschen und die 
Kategorie wieder freischalten für "ohne Passwort" ?

ThomasFinnern commented:
> setzte ich ein Passwort für eine Kategorie und gehe dann wieder in die Einstellungen, fehlt die Sprachverknüpfung..
Texte eingeführt
Das Kategorie-Passwort hat generell ein paar Fehler. Rücksetzen passier meist durch "Save" nach irgendeiner Änderung. Wenn user nicht Admin-Rechte hat, dann ist das Passwort weg. Wenn admin dann einmal in das Passwort Feld Klicken, sichern und es ist weg :-(

eumel1602 commented:
Es ist (für mich) schwer zu erkennen was geändert, oder welche Fehler behoben wurden. 

- Ich wollte eine Kategorie mit zwei Bildern darin -> löschen (Mülltonnensymbol) er fragt mich ob ich das wirklich will, und dann speichert er es ohne Meldung das die Kategorie nicht gelöscht wurde (sind ja zwei bilder drin) 

- Beim Passwortfeld, die Voreingetragenen Punkte: "...." iritieren, ich denke da sollte nichts drin sein. 

- [Weiterhin] Remove Passwort auf Ja und Speichern, nimmt das Passwort weiterhin nicht raus...

- Wenn ich in der FE JoomGalleryansicht bin kann ich zum Beispiel in einem Ordner Edit oder Delet klicken. Da kommt ein Fehler! (siehe die beiden Bilder:
![bildschirmabgriff](https://github.com/user-attachments/assets/1ee9cc31-44fa-45a4-aca5-6a5e65a98817)

![bildschirmabgriff2](https://github.com/user-attachments/assets/b52661c7-1b39-488c-958c-275e6a636acf)

Das selbe passiert bei den beiden Button (edit und delet) bei einem Bild im "FE" bei der Einzelbildansicht

-----------------

- Wenn ich Bilder hochgeladen habe, und theoretisch rausgehe aus der Maske und erneut reingehe in den bilder upload, **Zeigt es mir voreingestellt immer die am neusten erstelle Kategorie an**. **Dieses Feld sollte leer sein**, das man sich eine neue Kategorie auswählen muss! sonst kommt es vor, dass man vergisst die kategorie umzustellen und lädt die Bilder in die falsche Kategorie. 
Nehmen wir an es waren 100 Bilder oder mehr, kann man sie dann, weil wir noch keine Option für Massenverschiebung haben, nur einzeln in die neue Kategorie verschieben...

DANKE Thomas! Gruss

eumel1602 commented:
guten morgen! 
getestet auf deiner Beispielseite! http://www.joomgallery.tomfinnern.de/

+ Die "edit" und "delet buttons" sind verschwunden wie besprochen /
+ will ich bilder hochladen ist die kategorieauswahl leer, wie besprochen / 
+ Die Anzeige des PW Feldes, da Speichern, wiededer änderung und remove password funktioniert nun / 

- **ich kann im user upload keine Bilder mehr hochladen!!!** (fehler: Uploading the file DSC_0034.JPG with Uppy failed.)
- der versuch des loeschen eine Kategorie, mit bilder darin, bringt weiterhin keine Fehlermeldung

wenn ich ein bild in der einzelansicht anschaue kommt oben links teile von einem code. siehe bild:
![einzelbildansicht](https://github.com/user-attachments/assets/c7d8422f-85ae-4ab3-a9cb-3c6c00217f70)

ThomasFinnern commented:
> ich kann im user upload keine Bilder mehr hochladen!!! (fehler: Uploading the file DSC_0034.JPG with Uppy failed.)
Fixed
> wenn ich ein bild in der einzelansicht anschaue kommt oben links teile von einem code ("-->"). siehe bild:
Fixed
> der versuch des loeschen eine Kategorie, mit bilder darin, bringt weiterhin keine Fehlermeldung
Die Überprüfung ist fertiger code. Manuel muss sagen welche Änderung wo passt

eumel1602 commented:
**2 Dinge im User Panel**

Übersicht über die User images:
ich gebe als Filter z.b. die Zahl 12 ein (für bilder mit der "12")
nach dem Suchen möchte ich mit der "Clear" Taste die Eingabe wieder leeren. Dies funktioniert nicht...
Auch bei anderen Filterungen macht der "Clear" button nix

![1](https://github.com/user-attachments/assets/b41557bd-4317-49ea-b6b6-5e9c251dcf57)


Wenn ich bei den User Images auf die Spaltenüberschriften klicke, erwarte ich eine sortierung (umkehrung). da scheint aber nichts zu passieren ausser ein "anker" verlinkung die mich ganz nach oben im user panel bringt...

![2](https://github.com/user-attachments/assets/8635565a-c94a-4cb2-846e-586856846727)

MrMusic commented:
>3 Dinge im User Panel
>1. Übersicht über die User images:...
>2. Wenn ich bei den User Images auf die Spaltenüberschriften klicke, erwarte ich eine sortierung (umkehrung). da scheint aber nichts zu passieren ausser ein "anker" verlinkung die mich ganz nach oben im user panel bringt...

Das kann ich bestätigen. Liegt vermutlich an einem  Javascript Fehler in `searchtools.js: Uncaught TypeError: can't access property "append", this.theForm is null`
Ich persönlich halte aber die Liste der 'User Images' innerhalb des Control Panel grundsätzlich für wenig hilfreich. Diese Anzeige würde ich im Menüpunkt deaktivieren. Die Liste der letzten Bilder und letzten Kategorien finde ich sehr gut.

Bei meinen Tests habe ich noch folgendes festgestellt:

- Wenn in der Joomla Konfiguration die 'SEF URLs' eingeschaltet sind kommt jeweils ein `404 Page not found` beim Löschversuch oder beim Versuch den Status zu ändern (in der list view der Images und Categories).

**User Categories view:**
- Kleiner Wunsch: Es wäre schön, wenn das zugeordnete Thumbnail auch in der Liste angezeigt werden würde (analog wie im Backend).

**Edit Category:**
- Änderungen in den Custom Fields werden nicht gespeichert.

**User Images view:**
- Kleiner Wunsch: Es wäre schön, wenn 'Featured' und 'Hidden' auch in der Liste angezeigt werden würde (analog wie im Backend).
- Kleiner Wunsch: Es wäre schön, wenn die ungenehmigten (not approved) Images eine entsprechende Markierung hätten. Der Besitzer würde sich sonst wundern warum diese Bilder normal nicht im FE angezeigt werden. Alternativ könnte man im User Panel im 'Information' Block die Anzahl mit einer Verlinkung anzeigen (analog wie im Backend).

**Edit Image:**
- Änderungen in den Custom Fields werden nicht gespeichert.

**Ansonsten: Tolle Arbeit. Vielen Dank!**

ThomasFinnern commented:
I will look into it , wishes on christmas ? ;-)

MrMusic commented:
>I will look into it , wishes on christmas ? ;-)

Yes, christmas is coming soon!
I think about the wish list now... :wink: :love_letter: :joy:

ThomasFinnern commented:
The project is nearing its finish line. Therefore, a feature freeze is now in place. Only necessary bug fixes will be implemented. The project shall now be merged into a delivery version.

ThomasFinnern commented:
Nach grösseren Änderungen bite nochmal gründlich testen

Elfangor93 commented on joomgallery.xml:
```diff
@@ -1,29 +1,30 @@
-    <description>JoomGallery 4, a native gallery component for Joomla! 4.x, 5.x and 6.x</description>
-    <namespace path="src">Joomgallery\Component\Joomgallery</namespace>
-
+  <name>com_joomgallery</name>
```
Im joomgallery.xml stimmen die Einrückungen ganz und gar nicht mehr.

Elfangor93 commented on site/com_joomgallery/src/Model/UsercategoryModel.php:
```diff
@@ -0,0 +1,229 @@
+   */
+  public function getError($i = null, $toString = true)
+  {
+    // 2025.10.04: user images checkin does not find error message in model !manuel!
```
getError und setError soll man sowieso nicht mehr verwenden.
Deshalb habe ich die `$this->component->setError(...)`. Diese wird es auch nach der deprecation durch Joomla noch geben.

Elfangor93 commented on site/com_joomgallery/src/Model/UseruploadModel.php:
```diff
@@ -0,0 +1,376 @@
+   * @throws \Exception
+   * @since   4.2.0
+   */
+  protected function loadComponentParams(int $id = 0): void
```
Diese Methode ist hier überflüssig, da du das `JoomAdminModel` verwendest. Das macht bereits alles für dich.

Elfangor93 commented on site/com_joomgallery/src/View/Usercategories/HtmlView.php:
```diff
@@ -0,0 +1,237 @@
+      $this->isUserLoggedIn = false;
+    }
+
+//    // at least one category is needed for upload view
```
Can be deleted.

ThomasFinnern replied:
Kept for when root category is needed again

ThomasFinnern commented on site/com_joomgallery/src/Model/UseruploadModel.php:
```diff
@@ -0,0 +1,376 @@
+   * @throws  \Exception
+   * @since   4.2.0
+   */
+  public function __construct($config = [], $factory = null)
```
Removed

ThomasFinnern commented on site/com_joomgallery/src/Model/UseruploadModel.php:
```diff
@@ -0,0 +1,376 @@
+   * @return  AccessInterface   Object on success, false on failure.
+   * @since   4.2.0
+   */
+  public function getAcl(): AccessInterface
```
Removed

Elfangor93 commented on media/com_joomgallery/js/uppy/dist/uppy-uploader.js:
```diff
@@ -6130,11 +6130,16 @@ class jgProcessor extends _uppy_core__WEBPACK_IMPORTED_MODULE_2__["default"] {
-        let data  = JSON.parse(temp.data);
-        res = {success: true, status: response.status, message: split[0], messages: temp.messages, data: data};
+        // error by php detected: json part with ...{"success":.... expected
+        if (res.startsWith('<br />')) {
```
Yes it is either a JSON string or a HTML in case of a PHP error or warning. This HTML always starts with a '\n' in the backend. That's what I used to recognise the HTML output.

Elfangor93 commented on site/com_joomgallery/src/View/Usercategories/HtmlView.php:
```diff
@@ -0,0 +1,237 @@
+    parent::display($tpl);
+  }
+
+  // Todo: FiTh/Manuel use trait    ??? breadcrumbs/ router/sef ???
```
Actually this is a good idea. But I think I would do it later. Maybe create an issue for that.

Elfangor93 commented on site/com_joomgallery/src/Service/DefaultRouter.php:
```diff
@@ -125,6 +132,30 @@ public function __construct(SiteApplication $app, AbstractMenu $menu, ?CategoryF
     $imageform->setKey('id');
     $this->registerView($imageform);
```
This form views are then not needed anymore, right?

ThomasFinnern replied:
Removed

ThomasFinnern commented:
> Zudem würde ich die layouts editcat und editimg umbenennen zu edit. So macht es das Joomla Framework standardmässig und es ist immer einfacher das zu machen, was das Framework tut.

Habe es versucht, aber es gab dann den Fehler mit check com_joomgallery.userimage failed. ? Ebenso bei category

Elfangor93 commented on site/com_joomgallery/src/Service/DefaultRouter.php:
```diff
@@ -125,6 +132,30 @@ public function __construct(SiteApplication $app, AbstractMenu $menu, ?CategoryF
+    $userimage->setKey('id');
+    $this->registerView($userimage);
+
+    parent::__construct($app, $menu);
```
Das ist doppelt drin.
Mach bitte `parent::__construct($app, $menu);` nur einmal.

ThomasFinnern commented on site/com_joomgallery/src/Service/DefaultRouter.php:
```diff
@@ -497,4 +737,131 @@ private function getCategory($id, $available = null, $root = true): CategoryTabl
+   * @since   4.0.0
+   * @throws  \UnexpectedValueException
+   */
+  private function getImage($id, $available = null, $root = true): CategoryTable
```
removed

MrMusic commented:
Ist der PR soweit fertig, dass final getestet werden kann?
Oder wird es noch weitere Code-Änderungen geben?

Elfangor93 commented:
So, ich bin durch mit dem Code-Review.

Folgende Punkte sind für mich noch offen/diskutabel:

## Router: View-Beziehungen
Aktuell sind alle Views als Root-Views im Router definiert. Ich schlage jedoch folgene Beziehungen vor:

- Userpanel
- - Usercategories
- - Usercategory (verschachtelbar)
- - Userimages
- - - Userimage

Damit würden "schönere" URL's resultieren auch wenn keine explizite Menüstruktur erstellt wird.
Ichhabe dazu einen PR an @ThomasFinnern gemacht: https://github.com/ThomasFinnern/JoomGallery_fith/pull/2

## Upload view
Aktuell ist der Userupload eine eigene View mit eigenem Controller und Model. Sollten wir die Upload view nicht als Sublayout der Userimage view realisieren analog zum Backend?
@ThomasFinnern: Was ist deine Meinung dazu?

Elfangor93 commented:
> Datei annehmen, Grösse verändern, Daten extrahieren

Das macht ja der Upload-Service. Dazu brauchst du gar keine Model/View. Schlussendlich machst du im Userupload model nur das Speichern. Aber OK. Lassen wir das so.

ThomasFinnern commented:
>Ich weiß nicht ob das irgendwo dokumentiert ist... ?)

Ich habe da mal was vorbereitet ;-)
[Doku Vorschlag ](https://github.com/ThomasFinnern/JoomGallery_fith_dev/tree/main/.jg_dev_doc/jg_4x/site/userUpload)
Einzelne Dokumente (evtl. veraltet) zum ganzen
Das muss irgendwer in die allgemeine Doku überführen

eumel1602 commented:
> > Ich weiß nicht ob das irgendwo dokumentiert ist... ?)
> 
> Ich habe da mal was vorbereitet ;-) 

Stimmt, das habe ich unter 01 überlesen....! Super!
Ich glaube ich hatte schon zugesagt das in die vorhandene Doku auf der JG Webseite mal zu übernehmen! 

EDIT: willst du da noch mehr reinschreiben oder ist die auch "fertig"?

Elfangor93 commented:
> Kategorien ohne Bilder werden mit SEF Links im FE nicht nagezeigt. Platz in der ansicht bleibt leer (Bild)
Wenn SEF ausgeschaltet ist, werden die Kategorien ohne Bilder aber angezeigt...

Das deutet auf ein issue im Router hin.
@eumel1602 Kannst du mal testen, bei welchen Router Einstellungen genau der Fehler auftritt und bei welchen es funktioniert?

Hier eine Liste aller möglichen Einstell-Möglichkeiten des Routers:

**Joomla Global Configuration**
- Search Engine Friendly URLs to `Yes`
- Add Suffix to URL to `Yes`

**System - SEF Plugin Configuration**
- Enforce a suffic by redirect to `Yes`
- Strict handling of index.php to `No`
- Trailing slash for URLs to `No change`
- Strict Routing to `Yes` (important!)

**Joomla JoomGallery Configuration**
- Configuration inheritance to `Default`

**JoomGallery Global Configuration Set**
- Use real paths to `No`
- Routing to `Modern Router (default)`
- Remove IDs from URLs to `No`
- JG3 compatibility mode to `No`

Mein guess wäre, dass die Einstellung `Remove IDs from URLs` oder `Strict Routing` einen Einfluss auf diesen Error haben.

eumel1602 commented:
> @eumel1602 Kannst du mal testen, bei welchen Router Einstellungen genau der Fehler auftritt und bei welchen es funktioniert?

Willst du genau diese spezielle einstellung oder alles hin und her probieren? das ist schon relativ schwer genau abzudecken... Einstellung: A / B / C / D / E / F...... nur F an... und so weiter... ?
